R/rjournal_article.R

```diff
@@ -6,6 +6,30 @@
 #' @export
 rjournal_article <- function(..., citation_package = 'natbib') {
 
+  ## make sure that all files are correctly named according to the R journal
+  ## requirements https://journal.r-project.org/submissions.html
+
+  ## get name of Bib-file and Rmd-file
+  Rmd_file_name <- list.files(
+    path = ".", pattern = ".Rmd", include.dirs = FALSE)[1]
+  bib_file_name <- list.files(
+    path = ".", pattern = ".bib", include.dirs = FALSE)[1]
+
+  ## correct bibliography entry in Rmd-file
+  Rmd_file_text <- readLines(Rmd_file_name)
+  pat <- regexpr("(?<=\\\\bibliography{).+[^}]", Rmd_file_text, perl = TRUE)
+  regmatches(Rmd_file_text, pat) <- xfun::sans_ext(Rmd_file_name)
+  writeLines(text = Rmd_file_text, con =  Rmd_file_name)
+
+  ## correct bib-filename to match the name of the Rmd-file
+  file.rename(from = bib_file_name, to = paste0(xfun::sans_ext(Rmd_file_name), ".bib"))
+
+  ## create R file with executable R code as requested
+  knitr::purl(Rmd_file_name, documentation = 1)
+
+  ##remove objects
+  suppressWarnings(rm(Rmd_file_name, Rmd_file_text, bib_file_name))
```

Comment:
@RLumSK , I have think about this today.

IMO It seems a bit too much to do all that in the format function. At least this should be done with `rmarkdown` using one of the `pre_*` or `post_*` function in the format. This would let the function access easily the input file name and the metadata, which we want both. (the `bibliography` information is in the metadata field. 

So I agree that the required R file can be generated automatically from the Rmd file using `purl`. I think this could be done in the `post_knit` step where you can easily access the input file name to do that. 
* About the Rmd file name and bib filename, I think we can also only add to the `post_knit` step because in there we have access to both the input Rmd file name, and the metadata field. 
I think we could be less impactful on the user's files by checking and warning if they are not the same. I am not sure about renaming ourself the file of the user is a good idea. I would just remind the rule of the requirements (improving the documentation is also a good way to do that). 

What do you think ? 
Are you willing to try on that or do you want me to take the lead on this ? 


Note: 
* I think The bibliography should be added in the yaml header and not at the end of the Rmd file. That may have mislead you. As we can see it is correctly taken into account in the template
https://github.com/rstudio/rticles/blob/ec5e82b21c89ecb2d62ea935b1a3104fd0e48375/inst/rmarkdown/templates/rjournal_article/resources/template.tex#L15-L17

Reply:
@cderv Sorry for the late response I was away from the internet. I am not sure whether I understand the comments correctly. 

## BIB-file renaming

The renaming of the BIB-file is done because it is the requirement by the R Journal (https://journal.r-project.org/share/author-guide.pdf): 

> Create a minimal ‘filename.bib’ BibTeX file with only the entries you need, without unnecessary fields such as abstracts, and add \bibliography{filename} at the end of ‘filename.tex’, where filename is the same file name created above.

So the author does not have much choice. Whatever the name of the BIB-file is, in the end, the journal wants to have it named similar to the TEX-file. Since the template is created with `'rticles'` it should be, however, not a problem. Of course, we could at this to the documentation, but to me, it is somehow odd to modify things manually when I use `'rticles'` (same goes for checking and warning). 

Nevertheless, I can understand that the authors want to use the YAML-header to define the BIB-file because they are used to it and maybe someone uses the template and then copy & paste an already defined BIB-file or something like this. So my suggestion would be as follows: 

1. I add it to the YAML-header as requested. 
2. The moment the PDF is built, it creates a subfolder 'Submission' with all the files you would need using the correct naming scheme. So basically: Ready for submission; just in a different folder. This way, it does not touch the user's files but creates a new subfolder (writing on the hard drive, but `'rmarkdown'` is doing it anyway).

What do you think?

## R-file

> So I agree that the required R file can be generated automatically from the Rmd file using purl. I think this could be done in the post_knit step where you can easily access the input file name to do that.

Sorry, I did not get it. What do you mean with "input file name". The name has to be the same as the TEX-file and it is generated from the RMD-file. Why do we want to have an option for another file name? Probably I am overlooking something.

Reply:
> The moment the PDF is built, it creates a subfolder 'Submission' with all the files you would need using the correct naming scheme. So basically: Ready for submission; just in a different folder. This way, it does not touch the user's files but creates a new subfolder (writing on the hard drive, but 'rmarkdown' is doing it anyway).

I like this idea. 

To sum up the situation about the bib file.

* R journal requirement is to have the .bib file name like the .tex file.
* The `rjournal_article` has an example bib file called `RJreferences.bib` in its skeleton folder. It can't be renamed when copied to the user folder so it needs to be renamed afterwards manually or automatically. 
* Let's not that the `template.tex` already support the bibliography filename to be set in the yaml header (`bibliography: `) and I think it should be used instead of `\bibliography{RJreferences}` in the Rmd file. 
* The .tex filename will be the one 

What we can do : 
1. Check and warn the user if the bib filename provide in yaml is not the same as the tex name rendered, and let the user correct
2. Rename the bib file in the user's place - the original file would be renamed
3. Copy and rename the bib file in another submission folder - the original file would still be there but unuseful at the end

I would think 1 is enough as it is a good thing that a user is aware of the requirements of its submissions. Maybe 2 is what we should do if we want to do it ourself. 3 is in the middle but it can always have side effect when we begin to play with folder. 

@RLumSK, about the input file names, what I mean is that we can use `rmarkdown` included `pre_*` or `post_*` features in the `output_format` to have access to what we need, and do something safer that a `list.files` to get access to those names. 

For examples, if someone sets `output_file` in `rmarkdown::render`, all the files should be named accordingly. This value could be accessed in the post processor for example. Currently, your implementation won't allow that. 
I can move the logic where I think it is best if you want, once we have decided what is best to do.

Reply:
Ok to me, the best solution appears to be: 

* We do not touch the user's files, and the bibliography uses the YAML-header (with a default, but the user can do whatever wanted)
* We use the `post_*` feature to create the wanted files 
* We also use the `post_*` feature to define the output folder with default being `submission`. This way the user can even create something like 'Initial submission', 'Revision' whatsoever. 
* I have to update the documentation

If ok, I would implement it as detailed above and then we can see whether we need additional changes.

Reply:
Let's do it this way ! 

About those `pre_* ` and `post_*` steps in `?rmarkdown::output_format`, I think you have access to the Rmd file name in the `pre_knit` or `post_knit` step - maybe `purl` should be done there, then the file renamed according to `output_file` later on on the `post_processor` function.

For the output folder, not sure it is easy to make customisable. I let you try. If modifying the user's file is simpler, you'll tell me.

Thanks!

Reply:
@cderv I decided to hook this up on the post_processing function. I played with a couple of options, but the way it is now implemented was the only reasonable way. First I had the crazy idea that the output_dir can be actually defined as input, but then I realised that the argument `output_dir` is a dead-end in `rmarkdown::pdf_document()`. It calls an internal function named `pdf_pre_processor()` but this function does nothing with the argument. 

What I still don't like is the way I identify the figures before I copy them. However, for the moment I cannot come up with a better solution.

Reply:
Thanks a lot for having looked into this !

I feel that creating a submission folder may not be the safest way - you're right that the figures are not easy to deal with. From past experience with Rmarkdown, playing with the files and dirs could lead to unforseen issue. 

I think here we can go back one step to be more aligned with other **rticles** formats and what we want to support in this package

- renaming the tex file and the R file the same are ok to do because we create those anyway
- about the bibliography, I think it is the user responsability to provide one and we can't assume the default. If we can document in skeleton that it needs to be renamed, and also warn the user if not done. it seems enough. 
I know you said you find that odd to do the renaming yourself or have a warning, but as any other format, the bib file is provided by the user as one should fill it anyway - so it is one step to "save as" the file with another name. Also we can assume that the user should be familiar with the journal requirements on the bib expectation. 
At the end, **rticles**  is only here to help at its best, and if one renaming is left out for the user to do, it seems ok.
- For the figure, I'll check that it works as expected. 

Are you ok if I take it from here and modify your post processor step accordingly ?

Reply:
@cderv Sure, thanks, feel free to take it from here. 

For the bibliography (sorry again some arguments):

* Now the YAML-field `bibliography` defines everything and the user can do whatever they want with the file, and it is already consistent with the other formats in `'rticles'`, right?
* Creating the submission folder would be only unsafe if the folder already exists in the folder, for instance, for all users who updated the package and rebuild, therefore, the exception in the code. Still, even then, it would not overwrite other user files (but maybe I am overlooking something here). In either case (update or new template): If `output_dir` in the YAML-header does not exist, nothing happens. 
* There is another point to it: Before @phargarten2 raised the issue, frankly, I was not aware of it, because it does not feel logic (but maybe it is). What matters for LaTeX is that the entry in the TEX-file matches the BIB-file. Since the default was ok, it never crossed my mind that I should rename the file. Still, you have a strong point: the user should be aware of the author's guidelines, and I agree that `'rticles'` cannot and shall not take that away from the author. However, I also believe that it saves authors a lot of time if the format issues are handled all correctly (therefore my contribution). This would allow authors to focus on the essential aspects of the author's guidelines regarding science.
* A warning message is possible. But shouldn't a warning flag an exception? Here, the warning would become the standard message, shown to every user using the template and hitting the `Knit` button the first time. Except for one case that the authors read everything super carefully, and the first thing done was renaming the BIB-file **and** adjusting the TEX-template. 

I found the current way is elegant (it does not mean that the code couldn't be more sophisticated) and user friendly, and it reflects functionality I would have "expected" in the first place. 

However, you have the bigger picture in mind, and so of course, the final decision is all yours. Thank in either case for your work the professional support!

Reply:
I'll revisit to this PR after @cderv finishes the revision. Depending on the complexity of implementation, we might reconsider the decision on what to do for the bibliography file. Thank you @RLumSK for the contribution and discussion!

Reply:
I resolve this discussion and we can continue in comment feed.